### PR TITLE
Dshekhar/bfd 461 bfd commons

### PR DESCRIPTION
**Pull request by dshekhar18**
**JIRA Ticket:**
[BFD-461](https://jira.cms.gov/browse/BFD-461)

**User Story or Bug Summary:**

BFD V1: Implement an architectural approach to minimize redundancy

### What Does This PR Do?
 Creates a resuable package layer for Common Components which can be shared between BFD v1 and v2.
 
**BFD v1/v2 to take a middle approach and make best efforts to minimize redundancy:**

- Identify Common components and cross cutting concerns which are independent of HAPI data models and which can be safely abstracted for both v1 and v2.
- Have version specific Utiliy components which are tied to HAPI data models (stu3 and r4).
- Have FHIR/HAPI version specific Resource providers and Transformers.
- V1 is not largely" going to change. No new fields. Only production bug fixes and SLO adherence.

 

All functional and integration tests pass for lower environments and a test deploy is done before deploying to prod

### What Should Reviewers Watch For?
N/A

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then here's a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence: N/A.
    * Does this PR add any new software dependencies?  **No**.
    * Does this PR modify or invalidate any of our security controls? **No**.
    * Does this PR store or transmit data that was not stored or transmitted before? **No**.
* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR should not be merged unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons? **No**.

### What Needs to Be Merged and Deployed Before this PR?
N/A

### Submitter Checklist

I have gone through and verified that...:

* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] I have named this PR and its branch such that they'll be automatically be linked to the (most) relevant Jira issue, per: <https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html>.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. BCDA, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/Diagnosis.java [hunk 1]
```diff
@@ -125,14 +122,14 @@ public void setLabels(DiagnosisLabel label) {
    * Enumerates the various labels/tags that are used to distinguish between the various diagnoses
    * in a claim.
    */
-  static enum DiagnosisLabel {
-    /** Note: display text matches {@link ExDiagnosistype#PRINCIPAL}. */
+  public static enum DiagnosisLabel {
+    /** Note: display text matches ExDiagnosistype#PRINCIPAL. */
```
Comment (jzulim) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/Diagnosis.java:126]: I'm thinking we should stay consistent and utilize JS Doc {@link 

this comment applies to all comments which have had this tag removed.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/IcdCode.java [hunk 2]
```diff
@@ -1,17 +1,18 @@
-package gov.cms.bfd.server.war.stu3.providers;
+package gov.cms.bfd.server.war.commons;
 
+import gov.cms.bfd.server.war.stu3.providers.TransformerUtils;
```
Comment (jzulim) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/IcdCode.java:3]: 👍

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/Diagnosis.java (hunk 1 above)
Comment (jzulim) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/Diagnosis.java:126]: I would suggest we keep using JS Doc {@link to be consistent with other comments. The same comment applies to all changed comments in this PR.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/TransformerConstants.java [hunk 3]
```diff
@@ -91,164 +92,168 @@
...
    */
-  static final String CODED_IDENTIFIER_TYPE_UPIN_DISPLAY = "Unique Physician Identification Number";
+  public static final String CODED_IDENTIFIER_TYPE_UPIN_DISPLAY =
+      "Unique Physician Identification Number";
 
-  /** The Code System, as available at <a */
-  static final String CODING_SYSTEM_HL7_IDENTIFIER_TYPE =
+  /**
+   * The Code System, as available at <a
+   * href="https://terminology.hl7.org/1.0.0/CodeSystem-v2-0203.html">identifierType</a>.
```
Comment (jzulim) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/TransformerConstants.java:194]: 👍

File: apps/pom.xml [hunk 4]
```diff
@@ -561,6 +561,9 @@ checkJavaFormat
 						<goals>
 							<goal>jar</goal>
 						</goals>
+						<configuration>
+							<failOnError>false</failOnError>
```
Comment (jzulim) [inline apps/pom.xml:565]: Can you highlight the reason for this change? I imagine it's to ignore cleanup errors, but why is this needed all of a sudden?

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/CCWProcedure.java [hunk 5]
```diff
@@ -1,11 +1,11 @@
-package gov.cms.bfd.server.war.stu3.providers;
+package gov.cms.bfd.server.war.commons;
 
 import java.time.LocalDate;
 import java.util.Objects;
 import java.util.Optional;
 
 /** Models a ccwProcedure code entry in a claim. */
-final class CCWProcedure extends IcdCode {
+public final class CCWProcedure extends IcdCode {
```
Comment (jzulim) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/CCWProcedure.java:8]: Were you still wanting to rename this, CCWProcedure?

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/Diagnosis.java (hunk 1 above)
Comment (jzulim) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/Diagnosis.java:126]:
```suggestion
    /** Note: display text matches {@link ExDiagnosistype#PRINCIPAL}. */
```

Comment (jzulim) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/Diagnosis.java:126]: same applies to all comments where the JS Doc {@link tag was removed.

**Review by jzulim — CHANGES_REQUESTED** [review]
The most apparent shared components have been pulled into commons, nice. In your walkthrough you pointed out other areas that could be common'd in the future, but for a first pass, this looks reasonable to me. I made a suggestion to re-add the JS Doc {@link tags back into the comments for consistency sake. 

I pulled this PR down locally, built and hit all FHIR endpoints with success. Once the comments are fixed I think this is good to go.

p.s. Thanks for fixing the line breaks.,

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/Diagnosis.java (hunk 1 above)
Comment (dshekhar18) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/Diagnosis.java:126]: Some of the Java doc comments have been modified to be independent of HAPI FHIR data model version to be available as a common component for both BFD v1 and BFD v2.

File: apps/pom.xml (hunk 4 above)
Comment (dshekhar18) [reply apps/pom.xml:565]: To address Javadoc doclint. It was failing the build if all the doclint criteria's were not met. Not sure how it was being subdued before but it started cropping up. We can always come back and revisit this if needed.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/CCWProcedure.java (hunk 5 above)
Comment (dshekhar18) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/CCWProcedure.java:8]: Not in this iteration.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/Diagnosis.java (hunk 1 above)
Comment (dshekhar18) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/Diagnosis.java:126]: Responded above.

**Review by cbrunefearless — APPROVED** [review]
lgtm

**Review by jzulim — APPROVED** [review]
